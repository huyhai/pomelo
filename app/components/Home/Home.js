import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import { bindActionCreators } from 'redux';
import { View, Text, FlatList, Image, Button, SafeAreaView, ActivityIndicator } from 'react-native';
import { fetchFakeData, fetchFakeDataSuccess } from '../../services/general/generalActions'
import commonStyles from '../../common/style';
import styles from './style';

export class Home extends React.Component {

    static propTypes = {
        fetchFakeData: PropTypes.func.isRequired,
        fetchingData: PropTypes.bool
    };

    static defaultProps = {
        fetchingData: false
    };

    componentDidMount() {
        const { fetchFakeData } = this.props;
        fetchFakeData()
    }

    render() {
        const { listTransactions, fetchingData } = this.props;
        return(
            <SafeAreaView style={commonStyles.flex1}>
                <FlatList
                    style={commonStyles.flex1}
                    data={listTransactions}
                    renderItem={this.renderItem}
                    keyExtractor={this.keyExtractor}
                    removeClippedSubviews={true}
                    ItemSeparatorComponent={this.renderSeparator}
                />
                <Button
                    title="Update to REFUNDED"
                    onPress={this.clickUpdate}
                />
                { fetchingData && (
                    <View style={styles.loading}>
                        <ActivityIndicator size="large" />
                    </View>
                )}
            </SafeAreaView>
        );
    }

    keyExtractor = item => item.id;

    clickUpdate = () => {
        const { listTransactions, fetchFakeDataSuccess: updateDataWithNewStatus } = this.props;
        const data = listTransactions.map((item, index) => {
            item.state = 'REFUNDED';
            return item;
        });
        updateDataWithNewStatus(data)
    };

    renderItem = ({ item, index }) => {
        const { providerDisplayName, state, qr, amount, currency, id } = item;
        const { url } = qr;
        return (
            <View key={id} style={styles.itemParent}>
                <View style={styles.itemParentRow}>
                    <Image
                        style={styles.imgStyle}
                        source={{uri: url}}
                        resizeMode={"cover"}
                    />
                    <View style={{flex:1, padding: 10}}>
                        <Text>{`ID: ${id}`}</Text>
                        <Text>{`DisplayName: ${providerDisplayName}`}</Text>
                        <Text>{`Amount: ${amount}`}</Text>
                        <Text>{`Currency: ${currency}`}</Text>
                        <Text style={styles.statusText}>{`Status: ${state}`}</Text>
                    </View>

                </View>
            </View>

        )
    };

    renderSeparator = () => {
        return (
            <View style={styles.separatorLine}/>
        );
    };
}


export const mapStateToProps = state => {
    const { general } = state;
    return {
        listTransactions: general.listTransactions,
        fetchingData: general.fetchingData,
    }
};

export const mapDispatchToProps = dispatch =>
    bindActionCreators(
        {
            fetchFakeData,
            fetchFakeDataSuccess
        },
        dispatch
    );

export default connect(mapStateToProps, mapDispatchToProps)(Home);
