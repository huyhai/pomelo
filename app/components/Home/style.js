import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
    shipToText: {fontSize: 18},
    itemParent: {padding: 16, flex:1},
    itemParentRow: {flex: 1,justifyContent:"space-between", alignItems:"center", flexDirection: "row"},
    imgStyle: { height:85, width:85},
    separatorLine: {
        height: 1,
        width: "100%",
        backgroundColor: "darkgray",
    },
    statusText: {fontWeight: 'bold'},
    loading: {
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        justifyContent:"center", alignItems:"center"
    }
});
export default styles;
