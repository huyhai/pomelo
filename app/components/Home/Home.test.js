import React from 'react'
import renderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import { Home, mapDispatchToProps, mapStateToProps } from './Home'

const shallowRenderer = new ShallowRenderer();

const defaultProps = {
    fetchFakeData: jest.fn(),
    fetchFakeDataSuccess: jest.fn(),
};

it('renders correctly', () => {
  const tree = shallowRenderer.render(<Home {...defaultProps} />);
  expect(tree).toMatchSnapshot()
});

test('renders correctly componentDidMount', () => {
    const tree = renderer.create(<Home {...defaultProps} />);
    tree.getInstance().componentDidMount();
    expect(defaultProps.fetchFakeData).toBeCalled();
});

describe('handle Flatlist function', () => {
    const tree = renderer.create(<Home {...defaultProps} />);
    test('calls keyExtractor', () => {
        const key = tree.getInstance().keyExtractor({ id: 1 });
        expect(key).toBe(1)
    });

    test('check renderItem', () => {
        const props = {
            ...defaultProps
        };
        const SAMPLE_ITEM = { id: '123abc', providerDisplayName: 'abc', state: 'Ok', qr: {url: ''}, amount: 123, currency: 'vnd' };
        const tree = renderer.create(<Home {...props} />);
        const Item = tree.getInstance().renderItem({ item: SAMPLE_ITEM });
        expect(Item.key).toBe('123abc');
    });

    test('calls renderSeparator', () => {
        const key = tree.getInstance().renderSeparator();
        expect(key).not.toBeNull()
    })
});

test('renders correctly getTodaySaleEvent', () => {
    const props = {
        ...defaultProps,
        listTransactions: []
    };
    const tree = renderer.create(<Home {...props} />);
    tree.getInstance().clickUpdate();
    expect(defaultProps.fetchFakeDataSuccess).toHaveBeenCalled();
});

describe('mapDispatchToProps', () => {
    it('returns bound function to dispatch', () => {
        const dispatchProps = mapDispatchToProps();
        expect(dispatchProps).toEqual({
            fetchFakeData: expect.any(Function),
            fetchFakeDataSuccess: expect.any(Function)
        })
    })
});

describe('mapStateToProps', () => {
    it('state for language', () => {
        const initialState = { general: { listTransactions: null, fetchingData: false } };
        expect(mapStateToProps(initialState).listTransactions).toBeNull();
        expect(mapStateToProps(initialState).fetchingData).toBeFalsy();
    })
});
