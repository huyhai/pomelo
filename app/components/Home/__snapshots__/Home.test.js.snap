// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<SafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <FlatList
    ItemSeparatorComponent={[Function]}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    style={
      Object {
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
  <Button
    onPress={[Function]}
    title="Update to REFUNDED"
  />
</SafeAreaView>
`;
